Added Adafactor

hazemessamm opened the pull request:
## Proposed changes

Please include a description of the problem or feature this PR is addressing. If there is a corresponding issue, include the issue #.

I added implementation for Adafactor and I tested it on a toy network. it should be beneficial for anyone who wants to use an optimizer that does not take too much memory.

Inspired from: https://github.com/jettify/pytorch-optimizer/blob/master/torch_optimizer/adafactor.py

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

awni commented:
Very cool!

awni commented on python/mlx/optimizers.py:
```diff
@@ -498,3 +497,155 @@ def apply_single(
+        self.relative_step = relative_step
+        self.warmup_init = warmup_init
+
+    def compute_rms(self, inputs):
+        return mx.sqrt(mx.mean(mx.square(inputs)))
+
+    def compute_learning_rate(self, step, rms):
+        relative_step_size = self.learning_rate
+        if self.relative_step:
+            min_step = 1e-6 * step if self.warmup_init else 1e-2
+            relative_step_size = mx.minimum(min_step, 1 / mx.sqrt(mx.array(step)))
+
+        parameter_scale = 1.0
+        if self.scale_parameter:
+            parameter_scale = mx.maximum(mx.array(self.eps[1]), rms)
+        return parameter_scale * relative_step_size
+
+    def approximate_exp_moving_avg(self, exp_avg_sq_row, exp_avg_sq_col):
```
These should probably be `_function_name` since they are implementation details.

awni commented on python/mlx/optimizers.py:
```diff
@@ -498,3 +497,155 @@ def apply_single(
+
+        if self.weight_decay != 0:
+            parameter += parameter * (-self.weight_decay * learning_rate)
+        return parameter + (-update)
```
Do `parameter - update` so it's one operation instead of  two

awni commented on python/mlx/optimizers.py:
```diff
@@ -498,3 +497,155 @@ def apply_single(
+        state["step"] = step
+        use_first_moment = self.beta_1 is not None
+
+        rms = state.get("rms", self.compute_rms(parameter))
```
This is supposed to be fixed for the duration of the optimization?

hazemessamm replied:
If I understood your question correctly then no it should not be fixed, I removed it from the state.

awni commented on python/mlx/optimizers.py:
```diff
@@ -498,3 +497,155 @@ def apply_single(
+        relative_step_size = self.learning_rate
+        if self.relative_step:
+            min_step = 1e-6 * step if self.warmup_init else 1e-2
+            relative_step_size = mx.minimum(min_step, 1 / mx.sqrt(mx.array(step)))
+
+        parameter_scale = 1.0
+        if self.scale_parameter:
+            parameter_scale = mx.maximum(mx.array(self.eps[1]), rms)
```
Btw you don't need to pre-cast scalars to an `mx.array`. In fact it's probably better to let them stay as scalars so that type promotion works more easily. For example:

```
>>> mx.maximum(1.0, mx.array(0.0, mx.float16))
array(1, dtype=float16)
>>> mx.maximum(mx.array(1.0), mx.array(0.0, mx.float16))
array(1, dtype=float32)
```

If someone waas usign say `bfloat16 parameters and wanted the optimizer state in `bfloat16` as well, it would work without needing to change much if we don't pre-cast scalars.

hazemessamm replied:
That's really cool to know! Thanks for explanation 👍🏽

awni commented on python/mlx/optimizers.py:
```diff
@@ -498,3 +497,155 @@ def apply_single(
+            update = mx.rsqrt(exp_avg_sq) * gradient
+
+        update = update / mx.maximum(
+            mx.array(1.0), self.compute_rms(update) / self.clip_threshold
```
Same comment regarding scalar here.

awni reviewed (changes requested):
Looks really nice! I left a few comments. Since this one is pretty involved it would be great to do one of two things:

1. Share a training log (toy task is fine, or you can use the [MNIST mlx example](https://github.com/ml-explore/mlx-examples/tree/main/mnist))
2. Add some tests, though I'm not sure on the best way to test this one 🤔 

Either one (or even better both) would be great!

hazemessamm commented:
Thanks for the review, I learned a lot from it, I added simple training script and added a simple training log that shows the loss and the accuracy for 5 epochs using Adafactor.

awni commented:
Thank you for the contribution! This looks great!

I removed the MNIST benchmark as it doesn't make sense to include in our package. But it's good that you verified it works. I added a minor test just to make sure state / dtypes / shape are working.

awni reviewed (approved):
🚀
